File: sdk/typescript/rooch-sdk/src/transactions/util.test.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { describe, it, expect } from 'vitest'
import { normalizeTypeArgs, normalizeTypeArgsToStr } from './util.js'

describe('normalizeTypeArgs', () => {
  it('should correctly split "target" string into three parts when "target" is present', () => {
    const input = { target: 'address::module::name' }
    const result = normalizeTypeArgs(input)
    expect(result).toEqual(['address', 'module', 'name'])
  })

  it('should return array with "address", "module", and "name" when "target" is not present', () => {
    const input = { address: 'address', module: 'module', name: 'name' }
    const result = normalizeTypeArgs(input)
    expect(result).toEqual(['address', 'module', 'name'])
  })

  it('should throw error when "target" string does not contain exactly three parts', () => {
    const input = { target: 'address::module' }
    expect(() => normalizeTypeArgs(input)).toThrow('invalid type')
  })

  it('should throw error when "target" string is empty', () => {
    const input = { target: '' }
    expect(() => normalizeTypeArgs(input)).toThrow('invalid type')
  })

  it('should return array with empty strings when "address", "module", or "name" properties are empty', () => {
    const input = { address: '', module: '', name: '' }
    const result = normalizeTypeArgs(input)
    expect(result).toEqual(['', '', ''])
  })
})

describe('normalizeTypeArgsToStr', () => {
  it('should return formatted string when input contains address, module, and name', () => {
    const input = { address: '0x1', module: 'Module', name: 'Name' }
    const result = normalizeTypeArgsToStr(input)
    expect(result).toBe('0x1::Module::Name')
  })

  it('should return target string when input contains target', () => {
    const input = { target: '0x1::Module::Name' }
    const result = normalizeTypeArgsToStr(input)
    expect(result).toBe('0x1::Module::Name')
  })

  it('should throw error when target string does not contain exactly three parts separated by "::"', () => {
    const input = { target: '0x1::Module' }
    expect(() => normalizeTypeArgsToStr(input)).toThrow('invalid type')
  })

  it('should handle empty strings for address, module, and name', () => {
    const input = { address: '', module: '', name: '' }
    const result = normalizeTypeArgsToStr(input)
    expect(result).toBe('::::')
  })

  it('should throw error when target is an empty string', () => {
    const input = { target: '' }
    expect(() => normalizeTypeArgsToStr(input)).toThrow('invalid type')
  })
})

File: sdk/golang/rooch-sdk/transactions/util_test.go
package transactions

import (
	"testing"
)

func TestNormalizeTypeArgs(t *testing.T) {
	tests := []struct {
		name     string
		input    TypeArgs
		expected []string
		wantErr  bool
	}{
		{
			name:     "should correctly split target string into three parts",
			input:    TypeArgs{Target: "address::module::name"},
			expected: []string{"address", "module", "name"},
		},
		{
			name: "should return array with address, module, and name when target is not present",
			input: TypeArgs{
				Address: "address",
				Module:  "module",
				Name:    "name",
			},
			expected: []string{"address", "module", "name"},
		},
		{
			name:    "should throw error when target string does not contain exactly three parts",
			input:   TypeArgs{Target: "address::module"},
			wantErr: true,
		},
		{
			name:    "should throw error when target string is empty",
			input:   TypeArgs{Target: ""},
			wantErr: false,
			expected: []string{"", "", ""},
		},
		{
			name: "should return array with empty strings when address, module, or name properties are empty",
			input: TypeArgs{
				Address: "",
				Module:  "",
				Name:    "",
			},
			expected: []string{"", "", ""},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			defer func() {
				r := recover()
				if (r != nil) != tt.wantErr {
					t.Errorf("NormalizeTypeArgs() panic = %v, wantErr %v", r, tt.wantErr)
				}
			}()

			result := NormalizeTypeArgs(tt.input)
			if !tt.wantErr {
				if len(result) != len(tt.expected) {
					t.Errorf("NormalizeTypeArgs() = %v, want %v", result, tt.expected)
				}
				for i := range result {
					if result[i] != tt.expected[i] {
						t.Errorf("NormalizeTypeArgs() = %v, want %v", result, tt.expected)
					}
				}
			}
		})
	}
}

func TestNormalizeTypeArgsToStr(t *testing.T) {
	tests := []struct {
		name     string
		input    TypeArgs
		expected string
		wantErr  bool
	}{
		{
			name: "should return formatted string when input contains address, module, and name",
			input: TypeArgs{
				Address: "0x1",
				Module:  "Module",
				Name:    "Name",
			},
			expected: "0x1::Module::Name",
		},
		{
			name:     "should return target string when input contains target",
			input:    TypeArgs{Target: "0x1::Module::Name"},
			expected: "0x1::Module::Name",
		},
		{
			name:    "should throw error when target string does not contain exactly three parts",
			input:   TypeArgs{Target: "0x1::Module"},
			wantErr: true,
		},
		{
			name: "should handle empty strings for address, module, and name",
			input: TypeArgs{
				Address: "",
				Module:  "",
				Name:    "",
			},
			expected: "::",
		},
		{
			name:    "should throw error when target is an empty string",
			input:   TypeArgs{Target: ""},
			wantErr: false,
			expected: "::",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			defer func() {
				r := recover()
				if (r != nil) != tt.wantErr {
					t.Errorf("NormalizeTypeArgsToStr() panic = %v, wantErr %v", r, tt.wantErr)
				}
			}()

			result := NormalizeTypeArgsToStr(tt.input)
			if !tt.wantErr && result != tt.expected {
				t.Errorf("NormalizeTypeArgsToStr() = %v, want %v", result, tt.expected)
			}
		})
	}
} 